proposals/42-human-readable-mutable-names.md

```diff
@@ -0,0 +1,258 @@
+# Configurable DNS resolvers for human-readable mutable names
+
+Authors: @lidel
+
+Initial PR: https://github.com/protocol/web3-dev-team/pull/42 <!-- Reference the PR first proposing this document. Oooh, self-reference! -->
+## Purpose &amp; impact 
+
+Improve the way mutable names work in our stack.
+
+- Empower developers with reliable primitive for publishing updates on human-readable names
+- Future-proof the way we do human readable names via DNS interop
+- Reduce privacy risks related to DNS leaks  in plaintext 
+
+#### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+We see CID, IPNS record, and DNSLink being de-facto standard for cryptographically-verifiable content-addressing and human-readable names on the distributed web. 
+
+There are three distinct types of addresses:
+  - immutable content:  `/ipfs/{cid}`
+  - mutable content
+      - cryptographic, IPNS record signed with a specific libp2p key: `/ipns/{libp2p-key-cid}` 
+      - human-readable, [DNSLink](https://dnslink.io): `/ipns/{dns-name}`
+        (which in turn, resolves to either an immutable or mutable content path from DNS TXT record)
+
+
+**There are known gaps in the mutable story:**
+- IPNS records are only published by the original author, which severely limits the utility of our mutable names
+    - It is not possible to pin/republish `/ipns/{libp2p-key-cid}` names by other peers, so when original publisher disappears, IPNS links no longer resolve and we see "IPNS rot". 
```

Comment:
This isn't _entirely_ true, from a protocol standpoint, at least. If the IPNS author sets an infinite "EOL" on an IPNS record, anyone can republish. However, IPNS records expire by-default to limit how "stale" of a record an attacker could return to a client.

Reply:
While protocol is mostly ok, IPNS rot is a mixture of implementation choices:

- The default is not an infinite EOL, so everything expires unless original publisher bumps the record over and over again
- Expired records are ignored during resolution, even if no valid one is present and expired ones are all we have
- Only original publisher can bump EOL and keep record available by republishing it